File: backend/src/Operum.Model/DTOs/Widgets/WidgetDto.cs
namespace Operum.Model.DTOs.Widgets
{
    public class WidgetSourceFieldDto
    {
        public string Purpose { get; set; } = string.Empty;
        public string FieldId { get; set; } = string.Empty;
        public string FieldName { get; set; } = string.Empty;
    }

    public class WidgetSourceDto
    {
        public string Id { get; set; } = string.Empty;
        // e.g. "Monthly Totals: Day, Amount".
        public string Name { get; set; } = string.Empty;
        public List<WidgetSourceFieldDto> Fields { get; set; } = [];
        public string TrackerId { get; set; } = string.Empty;
        public string TrackerName { get; set; } = string.Empty;
        public int Order { get; set; }
    }

    // Not scoped to any one dashboard; see DashboardWidgetDto for a placement's rendering.
    public class WidgetDto
    {
        public string Id { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public string? Description { get; set; }
        public string ResultType { get; set; } = string.Empty;
        public string Code { get; set; } = string.Empty;
        // Line/Bar only.
        public string? Grouping { get; set; }
        public bool MatchedValuesOnly { get; set; }
        // Goal widgets only, in the value field's own string format.
        public string? GoalTarget { get; set; }
        // Goal widgets only. Null reads as HigherIsBetter.
        public string? GoalDirection { get; set; }
        public List<WidgetSourceDto> Sources { get; set; } = [];
    }
}

File: frontend/src/features/widgets/types/WidgetDto.ts
import { CreateAnalyticFieldDto } from "../../analytics/types/requests/CreateAnalyticDto";

export interface WidgetSourceFieldDto {
    purpose: string;
    fieldId: string;
    fieldName: string;
}

export interface WidgetSourceDto {
    id: string;
    /** The widget's definition read through this source's fields, e.g. "Monthly Totals: Day, Amount". */
    name: string;
    fields: WidgetSourceFieldDto[];
    trackerId: string;
    trackerName: string;
    order: number;
}

/** Not scoped to any dashboard; see DashboardWidgetDto for how a placement of this renders on a board. */
export interface WidgetDto {
    id: string;
    name: string;
    description?: string;
    resultType: string;
    code: string;
    /** Line/Bar only: how the axis field is bucketed before the code aggregates it. */
    grouping?: string;
    /** Combined charts only: restricts to x-axis values shared by every source. */
    matchedValuesOnly: boolean;
    /** Goal widgets only: the target, in the value field's format (a number, or hh:mm:ss for a duration). */
    goalTarget?: string;
    /** Goal widgets only: a GoalDirection value. Absent behaves as HigherIsBetter. */
    goalDirection?: string;
    sources: WidgetSourceDto[];
}

export interface EntriesWidgetDefinitionDto {
    id: string;
    name: string;
    trackerId: string;
    trackerName: string;
}

export interface CreateWidgetSourceRequestDto {
    trackerId: string;
    fields: CreateAnalyticFieldDto[];
}

export interface CreateWidgetDto {
    name?: string;
    description?: string;
    resultType: string;
    code: string;
    /** Line/Bar only: how the axis field is bucketed before the code aggregates it. */
    grouping?: string;
    matchedValuesOnly?: boolean;
    /** Goal widgets only, and required for them: a number or an hh:mm:ss duration. */
    goalTarget?: string;
    /** Goal widgets only: a GoalDirection value. Omitted behaves as HigherIsBetter. */
    goalDirection?: string;
    sources: CreateWidgetSourceRequestDto[];
}

/** Result type, code, sources, and field mapping are fixed at creation; create a new widget to change them. */
export interface UpdateWidgetDto {
    name?: string;
    description?: string;
    /** Goal widgets only. Omitted, the current target is kept. */
    goalTarget?: string;
    /** Goal widgets only. Omitted, the current direction is kept. */
    goalDirection?: string;
}

export interface CreateEntriesWidgetDto {
    trackerId: string;
    name?: string;
}

export interface UpdateEntriesWidgetDto {
    name?: string;
}
